Atm driver and fields design

bartgol opened the pull request:
This PR introduces some work related to the driver and fields structures. Some highlights:

 - added some utility strucutres: any class (mimicking std::any from c++17). a parameter list (same interface as Teuchos in Trilinos; in case we have Trilinos as a dependency in the future, we can switch), factory abstract class (following a common design pattern).
 - formalized the field and field-related structures; includes field structures (Field, FieldHeader, FieldIdentifier), some restriction on field creation/manipulation (e.g., cannot create a field until all dimensions are known), a field repository to store fields.
 - written some structures needed to handle parallel atm process splitting; namely, we now have an AtmosphereProcessGroup class, which can execute the stored processes in sequential or parallel mode. Right now, onl sequential is supported; some extra structures are needed to handle parallel (MPI) splitting, but we first need to write an interface for the grid structure (and grid remapping) before we can seriously thinking about how these MPI syncing/scattering classes will look like.

Note to parametrization writers: you can assume that the AtmosphereProcess interface won't change dramatically. The things that are more likely to change are the inputs to the initialize/run/finalize methods. Most likely, the initialize method will have additional inputs (e.g., some data related to number of elements/columns/tracers...), and the run method may receive some information related to the current time step (e.g., dt, t, ...). I don't foresee the finalize method receiving inputs.
When it's time to insert your parametrization in scream, write an interface class that inherits from AtmosphereProcess, and implement the purely virtual methods.

Feedback on what kind of information parametrizations will need to be init-ed/run is very welcome, since it will affect the signature of those methods.

ambrad commented on components/scream/src/share/field/field.hpp:
```diff
@@ -0,0 +1,172 @@
+
+  // Constructor(s)
+  Field () = delete;
+  Field (const identifier_type& id);
```
Single-arg ctors should be explicit.

ambrad commented on components/scream/src/share/field/field.hpp:
```diff
@@ -0,0 +1,172 @@
+
+// A field should be composed of metadata info (the header) and a pointer to the view
+
+template<typename DataType, typename MemSpace, typename MemManagement>
```
MemManagement is I think actually MemoryTraits.

ambrad commented on components/scream/src/share/field/field.hpp:
```diff
@@ -0,0 +1,172 @@
+  // Constructor(s)
+  Field () = delete;
+  Field (const identifier_type& id);
+  Field (std::shared_ptr<FieldHeader> fh, view_type view);
```
Not a big deal, but `const std::shared_ptr<T>&` saves needless smart-pointer bookkeeping.

ambrad commented on components/scream/src/share/field/field.hpp:
```diff
@@ -0,0 +1,172 @@
+        header_type& get_header ()       { return *m_header; }
+        view_type    get_view   () const { return  m_view;   }
+
+  std::shared_ptr<header_type> get_header_ptr () const { return m_header; }
```
Somewhat more important is to return `const sp<T>&` rather than `sp<T>` to save bookkeeping. Otherwise every caller has to be sure to grab the sp once explicitly and then deref that copy; calling this function repeatedly would have overhead.

ambrad commented on components/scream/src/share/scream_types.hpp:
```diff
@@ -1,18 +1,63 @@
+template <typename DataType, typename... Properties>
+using HostView = ViewType<DataType, HostMemSpace, Properties...>;
+template <typename DataType, typename... Properties>
+using ExecView = ViewType<DataType, ExecMemSpace, Properties...>;
```
We need to plan for mixed execution spaces, such as OpenMP/Cuda. In P3, we're avoiding these top-level type aliases in favor of type dictionaries templated on the device. So all of these top-level aliases go into a type dictionary.

bartgol replied:
Sounds reasonable. I mostly copied and pasted HOMMEXX stuff, but I guess hybrid parallelism is going to happen in the lifetime of scream.

ambrad commented on components/scream/src/share/scream_types.hpp:
```diff
@@ -1,18 +1,63 @@
+
+// Managed/Unmanaged view
+template <typename DataType, typename... Properties>
+using ViewManaged = ViewType<DataType, Properties..., MemoryManaged>;
```
scream_kokkos_meta.hpp provides an Unmanaged decorator that is more general than this in that it permits other memory traits.

ambrad replied:
(By the way, is it OK to merge the trivial PR #31 to clean up this decorator?)

bartgol replied:
Well, that one is a decorator, a type dependent on another type. If you use unmanged to view another view, then the decorator is perfect. But there will be places (e.g., surface coupling) where we will create unmanaged views from f90 pointers. In those areas, we won't have a reference managed view to deduce the unamanged templates arguments from.

But yes, in the FR I should probably use the unmanaged decorator to get an unmanaged view from a managed one.

ambrad commented on components/scream/src/share/mpi/scream_comm.hpp:
```diff
@@ -0,0 +1,51 @@
+  //   2) the call is collective on both the stored and input comm's
+  void reset_mpi_comm (MPI_Comm new_mpi_comm);
+
+  bool root () const { return m_rank==0; }
```
I think we want root() to return the root rank and am_root to return rank == root().

bartgol replied:
Well, isn't the root rank always 0? But yes, the check method can be more verbose, like `am_root`.

ambrad commented on components/scream/src/share/error_defs.hpp:
```diff
@@ -1,6 +1,8 @@
 #ifndef SCREAM_ERROR_DEFS_HPP
 #define SCREAM_ERROR_DEFS_HPP
 
+#include <string>
```
Sooner rather than later we probably want assertion and throw macros that take a comm, so the output can include the rank. Doesn't have to be this PR, though.

bartgol replied:
Yes, at some point we should have a PR focused on debug/assert/exceptions.

ambrad commented on components/scream/src/share/field/field_repository.hpp:
```diff
@@ -0,0 +1,76 @@
+ /*
+  *  A database for all the persistent fields needed in an atm time step
+  *  We template a field repository over the memory space. The data type of the underlying view
+  *  will *always* be Real*, so that we can store all the fields in one place (regardless of
```
Are you considering vectorization and, potentially, AD?

With vectorization, Real is OK b/c it can be the scalarized field. But, importantly for things like `reshape`, we have to be aware that the scalarized View will have padding if the pack size does not divide the number of levels. So, in summary, Real is OK here, but think through whether anything has incorrectly assumed no padding.

SFad-type AD I'm pretty sure scalarizes w/o an issue, as well. But you might check that. (I think it's reasonable to disallow DFad-type AD.)

ambrad replied:
I'm not suggesting a chance since SFad should be fine with Real: just asking whether padding has been considered in the design.

ambrad commented:
Would you explain the decisions behind Factory? On the surface, I see a few issues:

1. create() can create multiple objects with the same key. Is that desired? Should that be controlled? The fact that Factory::instance() makes object creation available to literally anyone means object creation is, on the surface, uncontrolled.
2. Factory::instance() means that there will be an arbitrary number of singletons. Basically, anything that inherits from Factory is given license to be a singleton. IMO, moving from uncontrolled singletons to one singleton, Context, to manage things in Hommexx was a good step.
3. A Factory is sometimes used to hide an impl. But in this case, atmosphere_process_registration.cpp `#includes` a bunch of headers containing impl details, anyway. I would argue that these headers should hide impl details by using the pImpl pattern (each process is very high level, so pImpl's deref cost doesn't matter), and the Factory itself doesn't seem necessary; each process can have a static `create` method. Note here that `create` needn't be exposed to just anyone; it can be isolated to some top-level object-broker (such as a Context).

ambrad commented:
Looks good. My main concerns are control of singletons, whether Real Fields accounts for vectorization (padding) and AD, and whether multiple execution spaces are supported. Anything I wrote beyond these three issues are opinions only and may be ignored w/o comment.

bartgol commented:
I was thinking about the factory myself. It's not really needed, after all, since atm processes are not created here and there, but only once, in a specific place (the AtmosphereProcessGroup class). At the beginning it seemed neat, but then there were some issues when registering products in a TU and using them in another TU (i.e., didn't work), so I had to introduce the registration file.

So no, I have no reason to keep the factory approach versus a hand-rolled `if(name=="bla"){} else if(name=="bla2"){}...` version. Both designs are mildly unsatisfactory to me, since high level classes/files need to be aware of all types, but there's no way around it. Perhaps without factory the code looks easier.

As for padding, the code does not keep that into account. I have to think about a clean way of doing it. I obviously don't want to have one FR per scalar type (for instance, two different process may use a vectorized field differently: one vectorized, the other not vectorized); on the other hand, I don't want to insert information about padding/vectorization in the identifier (it's an impl detail). Perhaps I should handle it in the view allocation process, but I don't have a clean solution in mind yet.

Finally, for execution spaces, I need to think of what's the friendliest interface for that. I think we can get away with a 'DeviceExecSpace' and 'HostExecSpace' (that is, just two execution spaces). I don't think we need to support more than one execution space on host. If we agree on that, then I think we can have a traits-like templated class where we can define all our view types.

Mostly, I wanted this branch to go in before it grew even larger. Some structure will change some detail, but the general ideas should not change much.

bartgol commented:
I'm demoting this PR to WIP, since accommodating padding in the fields framework is requiring a little more work than I thought. It's going to happen, but I have to do some modifications, and I don't want this to go in master until it's done.

bartgol commented:
I changed the field class quite a bit. Most notably, Field is now _always_ storing 1d views, and the first template parameter is now simply the scalar type. There was no need to store the full N-dimensional data type, since we didn't really use it inside the class, and it didn't add any real value.

I added support for padding, in case one wants to use the field with Pack types. This can complicate the life of the customers of the field that were not interested in Packs, since now they need to be aware of a change in the stored view dimensions. There are ways for a user to find out if this happened. I will add more doc in the next days to help guide the user.

@ambrad @jgfouca , if you may want to check the last mods (mainly the way Field looks after the last commit). The travis build fails because the atm driver unit test is not registering any field in the driver's field repo, and that's considered an error by the field repo (makes sense in normal runs, but it hurts now that no atm process is fully implemented). It should be fixed as soon as one atm process has a valid impl (it will probably be the surf coupling or the dynamics).

bartgol commented:
I incorporated some changes suggeste by the review. I'll leave the PR up foor further review, in case you want to. If no comments come, I will merge on Monday morning.

bartgol commented:
I pushed a couple of commits to fix code quality concerns of Codacity. The failing checks refer to code coverage, which will improve when the driver design proceeds. I take responsibility of adding tests to cover code paths, or to remove functions not actually needed. Hence, I will now merge.
